prow/github/client.go

```diff
@@ -1568,6 +1569,51 @@ func (c *client) GetBranches(org, repo string, onlyProtected bool) ([]Branch, er
 	return branches, nil
 }
 
+// GetBranchProtection returns current protection object for the branch
+//
+// See https://developer.github.com/v3/repos/branches/#get-branch-protection
+func (c *client) GetBranchProtection(org, repo, branch string) (*BranchProtection, error) {
+	c.log("GetBranchProtection", org, repo, branch)
+	code, body, err := c.requestRaw(&request{
+		method: http.MethodGet,
+		path:   fmt.Sprintf("/repos/%s/%s/branches/%s/protection", org, repo, branch),
+		// GitHub returns 404 for this call if either:
+		// - The branch is not protected
+		// - The access token used does not have sufficient privileges
+		// We therefore need to introspect the response body.
+		exitCodes: []int{200, 404},
+	})
+
+	switch {
+	case err != nil:
+		return nil, err
+	case code == 200:
+		bp := &BranchProtection{}
+		if err := json.Unmarshal(body, bp); err != nil {
+			return nil, err
+		}
+		return bp, nil
+	case code == 404:
+		// continue
+	default:
+		return nil, fmt.Errorf("unexpected status code: %v", code)
+	}
+
+	ge := &githubError{}
+	if err := json.Unmarshal(body, ge); err != nil {
+		return nil, err
+	}
+
+	// If the error was because the branch is not protected, we return a
+	// nil pointer to indicate this.
+	if ge.Message == "Branch not protected" {
```

Comment:
Relying on error strings is generally ill advised as it is brittle. Are you sure this is worth it?

Reply:
Thank you for the review. The github API for getting branch protection returns 404 for both a) unprotected branch, and b) authorisation error (perhaps for other `Not Found` errors as well). The only way to distinguish these seems to be looking at the error message body, which we need since `branchprotector` does not consider the former, i.e. unprotected branch, an error.